File: cpp/microbench/json_example/json_example.cpp
//
// Created by Ravil Galiev on 25.07.2023.
//

#ifndef SETBENCH_JSON_EXAMPLE_CPP
#define SETBENCH_JSON_EXAMPLE_CPP

#include <fstream>
#include <iostream>
#include "json/single_include/nlohmann/json.hpp"

#include "workloads/args_generators/impls/default_args_generator.h"
#include "workloads/thread_loops/impls/default_thread_loop.h"
#include "workloads/bench_parameters.h"

#include "workloads/distributions/builders/skewed_uniform_distribution_builder.h"

typedef long long ll;

ArgsGeneratorBuilder* getDefaultArgsGeneratorBuilder() {
    return (new DefaultArgsGeneratorBuilder())
        ->setDistributionBuilder((new ZipfianDistributionBuilder())->setAlpha(1.0))
        ->setDataMapBuilder(new ArrayDataMapBuilder());
}

ArgsGeneratorBuilder* getTemporarySkewedArgsGeneratorBuilder() {
    return (new TemporarySkewedArgsGeneratorBuilder())
        ->setSetNumber(5)
        ->setHotTimes(new long long[5]{1, 2, 3, 4, 5})
        ->setRelaxTimes(new long long[5]{1, 2, 3, 4, 5})
        ->setHotSizeAndRatio(0, 0.1, 0.8)
        ->setHotSizeAndRatio(1, 0.2, 0.7)
        ->setHotSizeAndRatio(2, 0.3, 0.6)
        ->setHotSizeAndRatio(3, 0.4, 0.6)
        ->setHotSizeAndRatio(4, 0.5, 0.7)
        ->enableManualSettingSetBegins()
        ->setSetBegins(new double[5]{0, 0.1, 0.2, 0.3, 0.05});
}

ArgsGeneratorBuilder* getCreakersAndWaveArgsGeneratorBuilder() {
    return (new CreakersAndWaveArgsGeneratorBuilder())
        ->setCreakersRatio(0.2)
        ->setWaveSize(0.2)
        ->setCreakersSize(0.1)
        ->setDataMapBuilder(new IdDataMapBuilder());
}

ThreadLoopBuilder* getDefaultThreadLoopBuilder(ArgsGeneratorBuilder* argsGeneratorBuilder) {
    return (new DefaultThreadLoopBuilder())
        ->setInsRatio(0.1)
        ->setRemRatio(0.1)
        ->setRqRatio(0)
        ->setArgsGeneratorBuilder(argsGeneratorBuilder);
}

ThreadLoopBuilder* getTemporaryOperationThreadLoopBuilder(ArgsGeneratorBuilder *argsGeneratorBuilder) {
    return (new TemporaryOperationsThreadLoopBuilder())
        ->setStagesNumber(3)
        ->setStagesDurations(new size_t[3]{1000, 2000, 3000})
        ->setRatios(0, new RatioThreadLoopParameters(0.1, 0.1, 0))
        ->setRatios(1, new RatioThreadLoopParameters(0.2, 0.2, 0))
        ->setRatios(2, new RatioThreadLoopParameters(0.3, 0.3, 0))
        ->setArgsGeneratorBuilder(argsGeneratorBuilder);
}

Parameters* getCreacersAndWavePrefiller(size_t range,
                                        CreakersAndWaveArgsGeneratorBuilder* argsGeneratorBuilder) {
    CreakersAndWavePrefillArgsGeneratorBuilder* prefillArgsGeneratorBuilder =
        new CreakersAndWavePrefillArgsGeneratorBuilder(argsGeneratorBuilder);
    return (new Parameters())
        ->addThreadLoopBuilder((new PrefillInsertThreadLoopBuilder())
                                   ->setArgsGeneratorBuilder(prefillArgsGeneratorBuilder))
        ->setStopCondition(
            new OperationCounter(prefillArgsGeneratorBuilder->getPrefillLength(range)));
}

int main() {
    /**
     * The first step is the creation the BenchParameters class.
     */

    BenchParameters benchParameters;

    /**
     * Set the range of keys.
     */

    benchParameters.setRange(2048);

    /**
     * Create the Parameters class for benchmarking (test).
     */

    Parameters* test = new Parameters();

    /**
     * We will need to set the stop condition and workloads.
     *
     * First, let's create a stop condition: Timer with 10 second (10000 millis).
     */

    StopCondition* stopCondition = new Timer(10000);

    /**
     * Setup a workload.
     */

    /**
     * in addition to the DefaultArgsGeneratorBuilder,
     * TemporarySkewedArgsGeneratorBuilder and CreakersAndWaveArgsGeneratorBuilder are also
     * presented in the corresponding functions
     */
    ArgsGeneratorBuilder* argsGeneratorBuilder
                = getDefaultArgsGeneratorBuilder();
//                = getCreakersAndWaveArgsGeneratorBuilder();
//                = getTemporarySkewedArgsGeneratorBuilder();

            /**
             * in addition to the DefaultThreadLoopBuilder,
             * TemporaryOperationThreadLoopBuilder is also presented in the corresponding function
             */
    ThreadLoopBuilder* threadLoopBuilder
                = getDefaultThreadLoopBuilder(argsGeneratorBuilder);
//                = getTemporaryOperationThreadLoopBuilder(argsGeneratorBuilder);

    /**
     * now add the ThreadLoopBuilders (you can add several different)
     * to the parameter class indicating the number of threads.
     * You can also optionally specify the cores to which threads should bind (-1 without binding).
     */
    test->addThreadLoopBuilder(
                    threadLoopBuilder, 8,
                    new int[8]{-1, -1, 0, 0, 1, 2, 3, 3}
            )
            ->setStopCondition(stopCondition);

    benchParameters.setTest(test)
        .createDefaultPrefill();
    //        .setPrefill(getCreacersAndWavePrefiller(
    //            2048, (CreakersAndWaveArgsGeneratorBuilder*)argsGeneratorBuilder));

    std::cout << "to json\n";

    nlohmann::json json = benchParameters;

    std::ofstream out("json_example.txt");

    out << json.dump(4);

    std::cout << "end\n";
}

#endif //SETBENCH_JSON_EXAMPLE_CPP

File: java/src/contention/benchmark/json/JsonExample.java
package contention.benchmark.json;

import contention.benchmark.workload.BenchParameters;
import contention.benchmark.workload.Parameters;
import contention.benchmark.workload.args.generators.abstractions.ArgsGeneratorBuilder;
import contention.benchmark.workload.args.generators.builders.*;
import contention.benchmark.workload.data.map.builders.*;
import contention.benchmark.workload.distributions.builders.*;
import contention.benchmark.workload.thread.loops.builders.*;
import contention.benchmark.workload.stop.condition.StopCondition;
import contention.benchmark.workload.stop.condition.Timer;
import contention.benchmark.workload.thread.loops.abstractions.ThreadLoopBuilder;
import contention.benchmark.workload.thread.loops.parameters.RatioThreadLoopParameters;

import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;

public class JsonExample {
    public static ArgsGeneratorBuilder getDefaultArgsGeneratorBuilder() {
        return new DefaultArgsGeneratorBuilder()
                .setDistributionBuilder(
                        new ZipfianDistributionBuilder()
                                .setAlpha(1.0)
                )
                .setDataMapBuilder(
                        new ArrayDataMapBuilder()
                );
    }

    public static ArgsGeneratorBuilder getTemporarySkewedArgsGeneratorBuilder() {
        return new TemporarySkewedArgsGeneratorBuilder()
                .setSetNumber(5)
                .setHotTimes(new int[]{1, 2, 3, 4, 5})
                .setRelaxTimes(new int[]{1, 2, 3, 4, 5})
                .setHotSizeAndRatio(0, 0.1, 0.8)
                .setHotSizeAndRatio(1, 0.2, 0.7)
                .setHotSizeAndRatio(2, 0.3, 0.6)
                .setHotSizeAndRatio(3, 0.4, 0.6)
                .setHotSizeAndRatio(4, 0.5, 0.7);
    }

    public static ArgsGeneratorBuilder getCreakersAndWaveArgsGeneratorBuilder() {
        return new CreakersAndWaveArgsGeneratorBuilder()
                .setCreakersRatio(0.2)
                .setWaveSize(0.2)
                .setCreakersSize(0.1)
                .setDataMapBuilder(new ArrayDataMapBuilder());
    }


    public static ThreadLoopBuilder getDefaultThreadLoopBuilder(ArgsGeneratorBuilder argsGeneratorBuilder) {
        return new DefaultThreadLoopBuilder()
                .setInsertRatio(0.1)
                .setRemoveRatio(0.1)
                .setArgsGeneratorBuilder(argsGeneratorBuilder);
    }

    public static ThreadLoopBuilder getTemporaryOperationThreadLoopBuilder(ArgsGeneratorBuilder argsGeneratorBuilder) {
        return new TemporaryOperationsThreadLoopBuilder()
                .setStagesNumber(3)
                .setStagesDurations(new int[]{1000, 2000, 3000})
                .setRatios(0, new RatioThreadLoopParameters(0.1, 0.1))
                .setRatios(1, new RatioThreadLoopParameters(0.2, 0.2))
                .setRatios(2, new RatioThreadLoopParameters(0.3, 0.3))
                .setArgsGeneratorBuilder(argsGeneratorBuilder);
    }

    public static void main(String[] args) {
        /**
         * The first step is the creation the BenchParameters class.
         */

        BenchParameters benchParameters = new BenchParameters();

        /**
         * Set the range of keys.
         */

        benchParameters.setRange(2048);

        /**
         * Create the Parameters class for benchmarking (test).
         */

        Parameters test = new Parameters();

        /**
         * We will need to set the stop condition and workloads.
         *
         * First, let's create a stop condition: Timer with 10 second (10000 millis).
         */

        StopCondition stopCondition = new Timer(10000);
        test.setStopCondition(stopCondition);

        /**
         * Setup a workload.
         */

        /**
         * in addition to the DefaultArgsGeneratorBuilder,
         * TemporarySkewedArgsGeneratorBuilder and CreakersAndWaveArgsGeneratorBuilder are also presented
         * in the corresponding functions
         */
        ArgsGeneratorBuilder argsGeneratorBuilder = getDefaultArgsGeneratorBuilder();

        /**
         * in addition to the DefaultThreadLoopBuilder,
         * TemporaryOperationThreadLoopBuilder is also presented in the corresponding function
         */
        ThreadLoopBuilder threadLoopBuilder = getDefaultThreadLoopBuilder(argsGeneratorBuilder);

        /**
         * now add the ThreadLoopBuilders (you can add several different)
         * to the parameter class indicating the number of threads.
         * You can also optionally specify the cores to which threads should bind (-1 without binding).
         */
        test.addThreadLoopBuilder(threadLoopBuilder, 8)
                .setStopCondition(stopCondition);


        benchParameters.setTest(test)
                .createDefaultPrefill();

        String json = JsonConverter.toJson(benchParameters);

        try (PrintWriter out = new PrintWriter("example.json", StandardCharsets.UTF_8)) {
            out.print(json);
            System.out.println("Successfully written data to the file");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

